### DISCUSSION TITLE
[sqlcipher] I execute PRAGMA key = '', but it doesn't work in diesel

### DISCUSSION THREAD

**Discussion by 724198847** [diesel-rs/diesel#3058] [Q&A]
hello, There is a problem and asking for help, sorry for my pool English.

I execute "PRAGMA key = '{mykey}'" and "PRAGMA cipher_compatibility = 4;" in diesel to do cipher but it doesn't work. I query "PRAGMA key" and "PRAGMA cipher_compatibility" but get empty. However, I can set something else like busy_timeout and query busy_timeout.

At the same time, it works when I open the db in rusqlite directly. So I am wondering if I did anything wrong.

`impl CustomizeConnection<SqliteConnection, crate::Error> for DatabaseCustomizer {
    fn on_acquire(&self, conn: &mut SqliteConnection) -> Result<()> {
        
        conn.pragma_set_busy_timeout(self.config.busy_timeout)?;

        if let Some(key) = &self.maybe_encryption_key {
            println!("### to decryption: {}", key.to_owned());
            conn.execute("pragma cipher_migrate")?;
            conn.execute("PRAGMA cipher_compatibility = 4;")?;
            conn.execute(&format!("PRAGMA key='{}';", key))?;
        }

        let res = sql::<Text>(&"PRAGMA cipher_compatibility").load::<String>(conn);
        match res {
            Ok((re)) => println!("cipher_compatibility: {:?}", re),   // get  []
            Err(err) => println!("{}", err),
        }

        let res_key = sql::<Text>(&"PRAGMA key").load::<String>(conn);
        match res_key {
            Ok((re)) => println!("key: {:?}", re),  // get  []
            Err(err) => println!("{}", err),
        }
        Ok(())
    }
}`

Here is my code. The database is locked, because I cannot open it with DB Browser, even though I entry the true key. And I can get the connection but cannot execute any sql, get error "file is not a database". Obviously, there is something wrong occurred during decryption. How can I solve this problem？ 🥺

**Comment by weiznich** [comment]
It is generally really hard to answer such questions without having a fully reproducible minimal example. That written: As a guess I would say that you just linked the wrong database library. It seems like your diesel compiled binary may use an ordinary libsqlite3 version, while the rusqlite compiled binary uses a sqlite version based on sqlchipher.

**Reply by 724198847** [reply]
Thank you for your answer. But this is my dependencies, i don't think there is any other diesel's feature can solve this problem. 

`[dependencies]
r2d2 = "0.8.9"
rusqlite = {version = ">=0.8.0, <0.27.0", optional = true, features = ["bundled-sqlcipher", "hooks"]}
diesel = {version = "1.4.8", features = ["sqlite", "r2d2"]}
diesel_derives = {version = "1.4.1", features = ["sqlite"]}
diesel_migrations = {version = "1.4.0", features = ["sqlite"]}
scheduled-thread-pool = "0.2.5"`

**Reply by weiznich** [reply]
To repeat that again: I cannot really say why that's this way without having a reproducible example I can take and run on my computer. I would guess that this is somehow related to which libsqlite3 version you link in, but that's something I cannot tell for sure without having access to your development environment. So either you are able to provide a complete reproducible example (remember it must be reproducible on some other computer) or your figure out the issue on your own. 
This will likely be my last answer here, at least till a reproducible example is provided.
